### PR TITLE
ML: update .ml-state actions to support > 1 index

### PR DESCRIPTION
**Pull request by benwtrent**
For smoother upgrades in the future, and the ability to have more than one state index, certain actions needed to be adjusted so that having more than one state index would not break key features. 

This addresses some (not all) of the points in https://github.com/elastic/elasticsearch/issues/29938

Namely:
- Make sure that all get requests for state documents are replaced with ID query _search requests that ignore duplicates
- Make sure all deletes can handle multiple indices
- Make sure index template has a wildcard so new state indices get the template

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java [hunk 1]
```diff
@@ -702,7 +702,7 @@ protected Clock getClock() {
 
             try (XContentBuilder stateMapping = ElasticsearchMappings.stateMapping()) {
                 IndexTemplateMetaData stateTemplate = IndexTemplateMetaData.builder(AnomalyDetectorsIndex.jobStateIndexName())
-                        .patterns(Collections.singletonList(AnomalyDetectorsIndex.jobStateIndexName()))
+                        .patterns(Collections.singletonList(AnomalyDetectorsIndex.jobStateIndexName() + "*"))
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java]: Would it be better to rename `AnomalyDetectorsIndex.jobStateIndexName()` to `AnomalyDetectorsIndex.jobStateIndexPattern()`, returning `".ml-state*"` and always use that when matching state indices?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobDataDeleter.java [hunk 2]
```diff
@@ -50,27 +52,33 @@ public JobDataDeleter(Client client, String jobId) {
...
+        Set<String> indices = new HashSet<>();
+        indices.add(stateIndexName);
         for (ModelSnapshot modelSnapshot : modelSnapshots) {
-            for (String stateDocId : modelSnapshot.stateDocumentIds()) {
-                bulkRequestBuilder.add(client.prepareDelete(stateIndexName, ElasticsearchMappings.DOC_TYPE, stateDocId));
-            }
-
-            bulkRequestBuilder.add(client.prepareDelete(AnomalyDetectorsIndex.jobResultsAliasedName(modelSnapshot.getJobId()),
-                    ElasticsearchMappings.DOC_TYPE, ModelSnapshot.documentId(modelSnapshot)));
+            idsToDelete.addAll(modelSnapshot.stateDocumentIds());
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobDataDeleter.java]: I think the 6.x version of this method also adds legacy v5.4 doc IDs.  It's probably best to raise a separate PR for the 6.x backport so we can review the 6.x code separately.  There may be other places where 5.x legacy is handled differently in 6.x.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java (hunk 1 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java]: Potentially, I can go ahead and do that in this PR, I was worried about simply making sure that if it was an alias backed by more than one index, that we would OK.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobDataDeleter.java (hunk 2 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobDataDeleter.java]: Sounds good to me, handling these whacky IDs is tricky. In current, the way we write the IDs prevents us from having duplicate IDs in different Indexes, not sure that is the case in v5.

**Comment by benwtrent** [comment]
run the gradle build tests 1

**Comment by benwtrent** [comment]
run the gradle build tests 2

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java (hunk 1 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java]: I kept the old name and added a new one. Many places that write to the index and such should probably hit the index directly (sans wildcard) for now until we add the aliases.

**Review by droberts195 — APPROVED** [review]
LGTM

**Pull request merged**